File: 2019W31/PreppindataW31.py
import pandas as pd
import numpy as np

actlog = pd.read_excel(pd.ExcelFile(r"E:\PD wk 31 input.xlsx"),"Northern Customer Orders")
actlog['Status'] = actlog['Status'].str.title()

order = actlog.pivot_table(index=['Order','Customer','City'],
                                  columns='Status', values='Date', aggfunc=max)
order.reset_index(inplace=True)
order['time to send'] = order.apply(lambda x: (x['Sent']-x['Purchased'])/np.timedelta64(1,'D'), axis=1)
order['time to review'] = order.apply(lambda x: (x['Reviewed']-x['Sent'])/np.timedelta64(1,'D'), axis=1)
order['Meet KPI'] = order['time to send'].apply(lambda x: 1 if x<=3 else 0)

ans1 = order.groupby(['Customer'], as_index=False).agg({'time to send':'mean'} \
                    ).rename(columns={'time to send':'Avg Time to Send'})
ans2 = order.groupby(['Customer'], as_index=False).agg({'time to review':'mean'} \
                    ).rename(columns={'time to review':'Time to Review from Sending Order'}).dropna()
ans3 = order[pd.isna(order['Sent'])][['City','Purchased','Sent','Order','Customer']]
ans3['Order not sent']='Not Sent'
ans4 = order.groupby(['City'], as_index=False).agg({'Order':'count', 'Meet KPI':'sum'} \
                    ).rename(columns={'Order':'Orders per City','Meet KPI':'Time To Send KPI'})
ans4['% Orders Meeting 3 Day KPI'] = ans4['Time To Send KPI'] / ans4['Orders per City']*100

File: 2019W31/preppindataw31.R
library(readxl)
library(dplyr)
library(tidyr)
library(stringr)
library(lubridate)

order <- read_excel("E:/PD wk 31 input.xlsx", 
                            sheet = "Northern Customer Orders", 
                     col_types = c("numeric", "date", "text", "text", "text")) %>%
  mutate_at(vars('Status'), ~str_to_title(.)) %>%
  spread(`Status`,`Date`) %>%
  mutate('Time to Send' = time_length(`Sent`-`Purchased`, 'day'),
         'Time to Review from Sending Order' = time_length(`Reviewed`-`Sent`, 'day'),
         'Order Not Sent' = ifelse(is.na(`Sent`),'Not Sent','Sent'))

ans1 <- order %>%
  group_by(`Customer`) %>%
  summarise('Avg Time to Send' = mean(`Time to Send`))

ans2 <- order %>%
  group_by(`Customer`) %>%
  summarise('Time to Review from Sending Order' = mean(`Time to Review from Sending Order`,na.rm=TRUE)) %>%
  drop_na()

ans3 <- order %>%
  filter(`Order Not Sent`=='Not Sent') %>%
  select(c('City','Order Not Sent','Purchased','Sent','Order','Customer'))
  
ans4 <- order %>%
  group_by(`City`) %>%
  summarise('Orders per City'= n(),
         'Time To Send KPI'= sum(`Time to Send`<=3,na.rm=TRUE),
         '% Orders Meeting 3 Day KPI' = `Time To Send KPI`/`Orders per City`*100)

View(ans1)
View(ans2)
View(ans3)
View(ans4)